### PR TITLE
LoadNexusProcessed: root-level NXentry groups

### PR DESCRIPTION
**Pull request by ekapadi**
### Description of work

Mantid's interpretation of the NeXus format uses multiple root-level "NX_class: NXentry" groups to store multiple workspaces in a single NeXus HDF5 file.  By using additional root-level groups, such as "NX_class: NXcollection", metadata about a group of workspaces may be stored.  In addition, the NeXus format itself allows other, non-NeXus information to be stored in an HDF5 file, provided it avoids collisions with the NeXus naming system.  In combination, all of this led to the requirement solved by the current PR, which allows `LoadNexusProcessed` to function correctly in the presence of other root-level non-NXentry and non-NeXus groups.

#### Summary of work

This commit includes the following changes:

  * When `LoadNexusProcessed` is determining the number of workspaces in a NeXus HDF5 file, it now counts the number of root-level "NX_class: NXentry" groups.  Previously, it counted the number of root-level groups, assuming all were of "NX_class: NXentry".

#### EWM ref:
[EWM#7148](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20%28Change%20Management%29#action=com.ibm.team.workitem.viewWorkItem&id=7148)

### To test:

A new unit test has been added to test these changes: `LoadNexusProcessedTest::test_load_workspace_group_other_root_groups`.

In order to verify these changes "by hand" it's recommended to take a look at the _contents_ of the test-input file: `build/ExternalData/Testing/Data/UnitTest/WorkspaceGroup_other_groups.nxs` using the `h5dump` tool.  In addition to including twelve workspaces (, as `NX_class: NXentry` "mantid_workspace_nn" groups), this file adds a root-level `NX_class: NXcollection` "metadata" group, and another root-level non-NeXus group.  Following the form of the new unit test, verify that `LoadNexusProcessed` loads this file as a group workspace, and successfully _ignores_ the other root-level groups in the file.

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

**Comment by sf1919** [comment]
Please can you add a milestone to this PR? If you can add labels too that would be good.

**Comment by walshmm** [comment]
Building too much longer than verifying :sweat_smile: 

### Code Review

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://developer.mantidproject.org/
- [ ] Are the unit tests small and test the class in 
- [ ] If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.
- [ ] If there are changes in the release notes then do they describe the changes 
- [ ] Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

### Functional 

- [ ] Do changes function as described? Add comments below that describe the tests 


Grep of the test file listing the 12 expected workspaces
image



And a sample of the newly added metadata:
image



all Root level groups can be see here:
image
where there are the two additional groups that are not workspaces.
- [ ] Do the changes handle unexpected situations, e.g. bad 
- [ ] Has the relevant (user and developer) documentation been added/updated?

**Comment by walshmm** [comment]
The test written does confirm that in the presence of these other groups it will function and load the file.

**Review by walshmm — DISMISSED** [review]
Approved, please find a gatekeeper!

File: buildconfig/CMake/CppCheck_Suppressions.txt.in [hunk 1]
```diff
@@ -733,7 +733,7 @@ constVariableReference:${CMAKE_SOURCE_DIR}/Framework/DataHandling/src/SaveFullpr
 passedByValue:${CMAKE_SOURCE_DIR}/Framework/Crystal/src/SaveHKL.cpp:579
 passedByValue:${CMAKE_SOURCE_DIR}/Framework/Crystal/src/SaveHKL.cpp:580
 constVariableReference:${CMAKE_SOURCE_DIR}/Framework/Crystal/src/SaveHKL.cpp:158
-constVariableReference:${CMAKE_SOURCE_DIR}/Framework/DataHandling/src/LoadNexusProcessed.cpp:922
+constVariableReference:${CMAKE_SOURCE_DIR}/Framework/DataHandling/src/LoadNexusProcessed.cpp:923
```
Comment (peterfpeterson) [inline buildconfig/CMake/CppCheck_Suppressions.txt.in:736]: Please fix the suppression by adding the word `const` to the right line and deleting this line.

File: Framework/DataHandling/test/LoadNexusProcessedTest.h [hunk 2]
```diff
@@ -37,8 +42,14 @@
 
 #include <string>
 
+// *** DEBUG ***
+#include <strstream>
```
Comment (peterfpeterson) [inline Framework/DataHandling/test/LoadNexusProcessedTest.h:46]: Please remove unused import

File: Framework/DataHandling/test/LoadNexusProcessedTest.h [hunk 3]
```diff
@@ -13,6 +13,11 @@
 #include "MantidAPI/NumericAxis.h"
 #include "MantidAPI/WorkspaceGroup.h"
 #include "MantidAPI/WorkspaceHistory.h"
+
+// *** DEBUG ***
+#include "MantidDataHandling/H5Util.h"
```
Comment (peterfpeterson) [inline Framework/DataHandling/test/LoadNexusProcessedTest.h:18]: Please remove unused imports

File: Framework/DataHandling/src/LoadNexusProcessed.cpp [hunk 4]
```diff
@@ -373,28 +374,28 @@ Workspace_sptr LoadNexusProcessed::doAccelleratedMultiPeriodLoading(NXRoot &root
 void LoadNexusProcessed::execLoader() {
 
   API::Workspace_sptr tempWS;
-  int nWorkspaceEntries = 0;
+  size_t nWorkspaceEntries = 0;
```
Comment (peterfpeterson) [inline Framework/DataHandling/src/LoadNexusProcessed.cpp:377]: Not a problem in this case, but openmp on windows used to not work with unsigned integer types. I keep meaning to check if they are supported yet. That is a frequent cause of casting integer types in mantid.

**Review by peterfpeterson — CHANGES_REQUESTED** [review]

File: buildconfig/CMake/CppCheck_Suppressions.txt.in (hunk 1 above)
Comment (ekapadi) [reply buildconfig/CMake/CppCheck_Suppressions.txt.in:736]: I'll add the `const`.  **But, my problem with this was that that line has to do with a _modification_, so it's definitely _not_ `const` -- that's why I preferred the suppression!**

Comment (peterfpeterson) [reply buildconfig/CMake/CppCheck_Suppressions.txt.in:736]: My misunderstanding. Please move the supression to be inline cpp file. Things that are going to stay should get moved over. Things in the suppressions file are meant to be dealt with when people want to go weeding.

Comment (ekapadi) [reply buildconfig/CMake/CppCheck_Suppressions.txt.in:736]: Added the suppression `inline`.  I also checked the signature:  definitely should _not_ be `const`.

File: Framework/DataHandling/test/LoadNexusProcessedTest.h (hunk 3 above)
Comment (ekapadi) [reply Framework/DataHandling/test/LoadNexusProcessedTest.h:18]: Removed.  Thanks for catching this!

**Review by peterfpeterson — APPROVED** [review]

**Pull request merged**